detection_rules/misc.py

```diff
@@ -296,6 +299,127 @@ def parse_user_config():
+
+    test_file: Optional[Path] = None
+    unit_tests: Optional[UnitTest] = None
+    rule_validation: Optional[RuleValidation] = None
+
+    @cached_property
+    def all_tests(self):
+        """Get the list of all test names."""
+        return discover_tests()
+
+    def tests_by_patterns(self, *patterns: str) -> List[str]:
+        """Get the list of test names by patterns."""
+        tests = set()
+        for pattern in patterns:
+            tests.update(list(fnmatch.filter(self.all_tests, pattern)))
+        return sorted(tests)
+
+    @staticmethod
+    def parse_out_patterns(names: List[str]) -> (List[str], List[str]):
+        """Parse out test patterns from a list of test names."""
+        patterns = []
+        tests = []
+        for name in names:
+            if name.startswith('pattern:') and '*' in name:
+                patterns.append(name[len('pattern:'):])
+            else:
+                tests.append(name)
+        return patterns, tests
+
+    def get_test_names(self, formatted: bool = False) -> (List[str], List[str]):
+        """Get the list of test names to run."""
+        patterns_t, tests_t = self.parse_out_patterns(self.unit_tests.test_only or [])
+        patterns_b, tests_b = self.parse_out_patterns(self.unit_tests.bypass or [])
+        tests = tests_t + tests_b
+        patterns = patterns_t + patterns_b
+        unknowns = sorted(set(tests) - set(self.all_tests))
+        assert not unknowns, f'Unrecognized test names in config ({self.test_file}): {unknowns}'
+
+        combined_tests = sorted(set(tests + self.tests_by_patterns(*patterns)))
+
+        if self.unit_tests.test_only:
+            tests = combined_tests
+            skipped = [t for t in self.all_tests if t not in tests]
+        elif self.unit_tests.bypass:
+            tests = []
+            skipped = []
+            for test in self.all_tests:
+                if test not in combined_tests:
+                    tests.append(test)
+                else:
+                    skipped.append(test)
+        else:
+            tests = self.all_tests
+            skipped = []
+
+        if formatted:
+            def fmt_tests(lt) -> List[str]:
+                raw = [t.rsplit('.', maxsplit=2) for t in lt]
+                ft = []
+                for test in raw:
+                    path, clazz, method = test
+                    path = f'{path.replace(".", os.path.sep)}.py'
+                    ft.append('::'.join([path, clazz, method]))
+                return ft
+
+            return fmt_tests(tests), fmt_tests(skipped)
+        else:
+            return tests, skipped
+
+    def check_skip_by_rule_id(self, rule_id: str) -> bool:
+        """Check if a rule_id should be skipped."""
+        bypass = self.rule_validation.bypass
+        test_only = self.rule_validation.test_only
+        if not (bypass or test_only):
+            return False
+        return (bypass and rule_id in bypass) or (test_only and rule_id not in test_only)
```

Comment:
For readability, do we want to expand/refactor this?

e.g.

```python
       if self.rule_validation.bypass and rule_id in self.rule_validation.bypass:
            return True
        if self.rule_validation.test_only and rule_id not in self.rule_validation.test_only:
            return True
        return False
```

or even, close to as-is

```python
       bypass_tests = self.rule_validation.bypass
       test_only_tests = self.rule_validation.test_only
       if bypass_tests and rule_id in bypass_tests:
            return True
        if test_only_tests and rule_id not in test_only_tests:
            return True
        return False
```